flips/20210916-nft-metadata.md

```diff
@@ -0,0 +1,354 @@
+
+## Design Proposal
+
+The core of this proposal is to add the following interface and have the NFT resource implement it
+```
+pub resource interface ViewResolver {
+  pub fun getViews() : [Type]
+  pub fun resolveView(_ view:Type): AnyStruct
+}
+```
+
+We will also add default implementations for these methods so that no compatibility will be broken 
+for existing NFT solutions. The new NFT interface could look something like this: 
+```
+pub resource NFT: INFT, ViewResolver {
+    pub let id: UInt64
+    pub fun getViews(): [Type] {
+       return []
+    }
+    pub fun resolveView(_ type: Type): AnyStruct {
+        return nil
+    }
+}
+```
+
+We propose that all metadata be expressed in Cadence types. The `pub fun getViews(): [Type]` 
+method allows for discovery of the various types of metadata supported by the NFT by returning a `Type` 
+for each supported type of metadata. An instance of any supported type could then be retrieved by 
+calling the `pub fun resolveView(_ type: Type): AnyStruct` method passing the desired type in as 
+an argument.
+
+This proposal by no means precludes the use of member variable on the NFT resource itself. In fact,
+ideally all immutable data for an NFT is stored as immutable members of the NFT resource and views 
+are resolved and created by accessing those immutable members.
+
+### Supporting other resources besides NFTs
+In order to support any other resource and resource collections, like marketplaces and fungible tokens, the following new interface is proposed:
+```
+pub resource interface ViewResolverCollection {
+    pub fun borrowViewResolver(id: UInt64): &{ViewResolver}
+    pub fun getIDs(): [UInt64]
+}
+```
+
+This will allow any collection to be exposed as view resolvable.
+
+For single resources that are stored as is like a fungible token or a profile contract you can just implement the `ViewResolver` interface directly.
+
+### Converting between different views
+
+Standards will evolve and new views will be added. In order to make it easier for maintainers we propose a mechanism to register a converter from one View to another:
+
+```
+pub struct interface ViewConverter {
+    pub let to: Type
+    pub let from: Type
+
+    pub fun convert(_ value:AnyStruct): AnyStruct
+}
+```
+
+[The Dandy NFT contract in Find has an example on how to do this](https://github.com/findonflow/find/blob/main/contracts/Dandy.cdc#L328). It would also be possible to create a central repository of ViewConverters where you send in a ViewResolver interface to get and resolve views.
+
+ 
+### Basic Views
+
+The most basic example is having `String` be a view. Think of this as `toString()` in Java or a `Stringer` implementation in Go. It is a way to view your NFT as a string. 
+
+The bigger brother of `String` would be `Display`. Here is how you could display an NFT in a wallet, marketplace or other user-facing application:
+```
+pub struct Display{
+    pub let name: String
+    pub let thumbnail: String
```

Comment:
should `thumbnail` be a `URLWithHash` instead? (ie: a URL to an image and a hash to verify it's contents).

Reply:
I have no problem with that, but I think the correct place for this discussion is in the NFT repo where we suggest the Display struct.